FuncCallTrace/Assets/Src/Scripts/CodeExecutor.cs

```diff
@@ -0,0 +1,137 @@
+          _buffer.Clear();
+        }
+      } else {
+        _buffer.Append(c);
+      }
+    }
+  }
+
+  private const string _defaultModuleName = "Program";
+  private const float _singleStepWaitInSeconds = .3f;
+
+  private readonly GameManager _gameManager;
+  private readonly ConsoleWriter _consoleWriter;
+  private readonly Engine _engine = new Engine(SeedXLanguage.SeedPython, RunMode.Script);
+  private readonly HashSet<string> _excludeFuncNames = new HashSet<string> () {
+    // TODO: Ignore all built-in function calls
+    "print"
+  };
+
+  // If the execution is running. It's treated as running if the engine is running or paused.
+  public bool IsRunning => !_engine.IsStopped;
+
+  // A switch for the caller to stop the code execution when the coroutine is running.
+  public bool Stopping = false;
+
+  public CodeExecutor(GameManager gameManager) {
+    _gameManager = gameManager;
+    _consoleWriter = new ConsoleWriter(gameManager);
+    _engine.RedirectStdout(_consoleWriter);
+    _engine.Register(this);
+  }
+
+  // Runs a SeedLang script. Returns false if the executor is already running.
+  public bool Run(string source) {
+    if (IsRunning || string.IsNullOrEmpty(source)) {
+      return false;
+    }
+    _gameManager.StartCoroutine(RunProgram(source));
+    return true;
+  }
+
+  public void On(Event.SingleStep e, IVM vm) {
+    // Highlights the current line.
+    _gameManager.QueueHighlightCodeLineAndWait(e.Range.Start.Line, _singleStepWaitInSeconds);
+    vm.Pause();
+  }
+
+  public void On(Event.FuncCalled e, IVM vm) {
+    if (_excludeFuncNames.Contains(e.Name)) {
+      return;
+    }
+    var args = new List<string>();
+    foreach (var arg in e.Args) {
+      args.Add(arg.ToString());
+    }
+    string label = $"{e.Name}({string.Join(',', args)})";
+    _gameManager.QueueFuncCalled(label);
+  }
+
+  public void On(Event.FuncReturned e, IVM vm) {
+    if (_excludeFuncNames.Contains(e.Name)) {
+      return;
+    }
+    string label = $"return: {e.Result}";
+    _gameManager.QueueFuncReturned(label);
+  }
+
+  // The coroutine to execute the source code.
+  private IEnumerator RunProgram(string source) {
+    _gameManager.QueueProgramStarted();
+    var collection = new DiagnosticCollection();
+    if (_engine.Compile(source, _defaultModuleName, collection)) {
+      if (_engine.Run(collection)) {
+        yield return new UnityEngine.WaitUntil(() => _gameManager.IsActionQueueEmpty);
+        while (!Stopping && !_engine.IsStopped) {
+          if (_engine.Continue(collection)) {
+            yield return new UnityEngine.WaitUntil(() => _gameManager.IsActionQueueEmpty);
+          } else {
+            _gameManager.QueueOutputSeedLangDiagnostics(collection);
+          }
+        }
+        if (Stopping) {
+          _engine.Stop();
+          Stopping = false;
+        }
+        _gameManager.QueueHighlightCodeLineAndWait(-1, 0);
+        _gameManager.QueueOutputTextInfo("Done.");
+      }
```

Comment:
The _engine.Run() function might return false as well. Do we need handle it?

And the CodeExecutor class is almost the same for all examples, Is it possible to extract common login into a base class?

Reply:
Yes all existing example projects have this issue since I copied the code from one to another. I will create an issue to log this and fix them together.

These projects are example code of SeedLang. I think it's not good to extract common code too much since users are learning SeedLang and expecting to see most of the embedding logic in one single file. It's necessary to have duplicate code among these projects.